include/exadg/incompressible_navier_stokes/spatial_discretization/operators/weak_boundary_conditions.h

```diff
@@ -460,12 +460,13 @@ inline DEAL_II_ALWAYS_INLINE //
       dealii::Tensor<1, dim, dealii::VectorizedArray<Number>> h;
       if(variable_normal_vector == false)
       {
-        h = FunctionEvaluator<1, dim, Number>::value(bc, q_points, time);
+        h = FunctionEvaluator<1, dim, Number>::value(*bc, q_points, time);
       }
       else
       {
         auto normals_m = integrator.get_normal_vector(q);
-        h              = FunctionEvaluator<1, dim, Number>::value(bc, q_points, normals_m, time);
+        h              = FunctionEvaluator<1, dim, Number>::value(
+          *(std::dynamic_pointer_cast<FunctionWithNormal<dim>>(bc)), q_points, normals_m, time);
```

Comment:
This will be required more often in the code if `FunctionWithNormal` is used more often.

Reply:
Yes, but that should be fine as the function call is specialized for `FunctionWithNormal` with the `normals` parameter. So no actual polymorphism is present here and it is better to be explicit about this. It should be the caller's responisiblity to call a function with the correct parameters (in which case it cannot fail), instead of the function throwing an exception if a condition is not met inside it, in this case if the dynamic cast failed.

Looking at it now, I think we should even make this a `static_pointer_cast`, as the actual type is known to be a `FunctionWithNormal` at compile time so that this function can be called at all.

Reply:
I don't understand why one should know at compile time that we have a `FunctionWithNormal`. Looking at the `BoundaryDescriptor`, we only know that we have a pointer to `dealii::Function`, don't we? So I would argue we have a downcast here. 

Hence, if the user makes a mistake in the application.h file, the current code with `static_pointer_cast` could lead to undefined behavior, while a dynamic cast should give a seg-fault, right?

So I think we have to keep the dynamic_pointer_cast here.

Reply:
Yeah, I meant it because of the `if-else` condition that we have here. If that boolean variable can be set independently from the function's type, such a mistake can of course occur. That isn't safe in and of itself, but you're right that a `dynamic_pointer_cast` will catch that error eventually. So sure, I'll make that a dynamic cast.